sled-agent/zone-images/src/zone_manifest.rs

```diff
@@ -0,0 +1,823 @@
+                         will not bring up zones that mismatch";
+                        "data" => %result.display(),
+                    );
+                }
+            }
+            Err(error) => {
+                // This error most likely requires operator intervention -- if
+                // it happens, we'll continue to bring sled-agent up but reject
+                // all Omicron zone image lookups.
+                error!(
+                    log,
+                    "error reading zone manifest for boot disk, \
+                     will not bring up Omicron zones";
+                    "error" => InlineErrorChain::new(error),
+                );
+            }
+        }
+
+        if self.non_boot_disk_metadata.is_empty() {
+            warn!(
+                log,
+                "no non-boot zpools found, unable to verify consistency -- \
+                 this may be a hardware issue with the non-boot M.2"
+            );
+        }
+
+        for info in &self.non_boot_disk_metadata {
+            info.log_to(&log);
+        }
+    }
+}
+
+#[derive(Clone, Debug, PartialEq)]
+pub struct ZoneManifestNonBootInfo {
+    /// The name of the zpool.
+    pub zpool_name: ZpoolName,
+
+    /// The dataset directory.
+    pub dataset_dir: Utf8PathBuf,
+
+    /// The zone manifest path.
+    pub path: Utf8PathBuf,
+
+    /// The result of performing the read operation.
+    pub result: ZoneManifestNonBootResult,
+}
+
+impl ZoneManifestNonBootInfo {
+    pub(crate) fn new(
+        info: InstallMetadataNonBootInfo<OmicronZoneManifest>,
+    ) -> Self {
+        let result =
+            ZoneManifestNonBootResult::new(&info.dataset_dir, info.result);
+        Self {
+            zpool_name: info.zpool_name,
+            dataset_dir: info.dataset_dir,
+            path: info.path,
+            result,
+        }
+    }
+
+    pub(crate) fn log_to(&self, log: &slog::Logger) {
+        let log = log.new(o!(
+            "non_boot_zpool" => self.zpool_name.to_string(),
+            "non_boot_path" => self.path.to_string(),
+        ));
+        self.result.log_to(&log);
+    }
+}
+
+impl IdOrdItem for ZoneManifestNonBootInfo {
+    type Key<'a> = ZpoolName;
+
+    fn key(&self) -> Self::Key<'_> {
+        self.zpool_name
+    }
+
+    id_upcast!();
+}
+
+#[derive(Clone, Debug, PartialEq)]
+pub enum ZoneManifestNonBootResult {
+    /// The manifest is present and matches the value on the boot disk.
+    /// Information about individual zone hashes matching is stored in the
+    /// `ZoneManifestArtifactsResult`.
+    Matches(ZoneManifestArtifactsResult),
+
+    /// The manifest is absent -- this is an error case because it is expected
+    /// to be present.
+    NotFound,
+
+    /// A mismatch between the boot disk and the other disk was detected.
+    Mismatch(ZoneManifestNonBootMismatch),
+
+    /// An error occurred while reading the mupdate override info on this disk.
+    ReadError(InstallMetadataReadError),
+}
+
```

Comment:
Many of these types are duplicated with some minor modifications -- deliberately so, since there are subtle differences (e.g. the underlying `install_dataset_metadata` doesn't error out if the file is missing, but here we do).